neuralmonkey/decoders/transformer.py

```diff
@@ -161,33 +161,10 @@ def __init__(self,
         self.attention_combination_strategy = attention_combination_strategy
         self.n_heads_hier = n_heads_hier
 
-        self.encoder_states = [get_attention_states(e) for e in self.encoders]
-        self.encoder_masks = [get_attention_mask(e) for e in self.encoders]
-
-        if self.encoder_states:
-            self.dimension = (
-                self.encoder_states[0].get_shape()[2].value)  # type: int
-
-            for i, enc_states in enumerate(self.encoder_states):
-                enc_dim = enc_states.get_shape()[2].value
-                if enc_dim != self.dimension:
-                    raise ValueError(
-                        "Dimension of the {}-th encoder ({}) differs from the "
-                        "dimension of the first one ({})."
-                        .format(i, enc_dim, self.dimension))
-
-        elif not self.embedding_size:
-            raise ValueError("'embedding_size' must be specified when "
-                             "no encoders are provided")
-        else:
-            self.dimension = self.embedding_size
-
-        if not self.dimension:
-            raise ValueError("Decoder could not infer model dimension")
-
-        if self.embedding_size != self.dimension:
-            raise ValueError("Model dimension and input embedding size"
-                             "do not match")
+        self.encoder_states = lambda: [get_attention_states(e)
```

Comment:
Proč jsou to lambdy a ne `@tensor`?

Reply:
to je humus kvůli beam searchi. Každej autoregresivní dekodér musí mít tohle, aby se mu místo tohohle během beam searche mohl podstrčit tensor expandovaný do batche.

Když se podíváš na tensorflowovej beam search decoder, řešej tam podobný pekla, takže bych to nechal.
Lambda tu je kvůli laziness.
Kdyby to bylo @tensor, nedalo by se to v poděděný třídě předefinovat.